File: screenShotServer/sssEclipseProject/src/adshotrunner/campaigns/Campaign.java
package adshotrunner.campaigns;

import java.io.File;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Timestamp;
import java.text.SimpleDateFormat;
import java.time.Instant;
import java.util.Calendar;
import java.util.HashSet;
import java.util.LinkedHashSet;
import java.util.Set;
import java.util.TreeSet;

import adshotrunner.errors.AdShotRunnerException;
import adshotrunner.powerpoint.CampaignPowerPoint;
import adshotrunner.powerpoint.PowerPointBackground;
import adshotrunner.system.ASRProperties;
import adshotrunner.utilities.ASRDatabase;
import adshotrunner.utilities.FileStorageClient;

/**
 * Class to retrieve and manipulate Campaigns.
 *
 * This class mirrors the PHP and Javascript AdShot classes.
 * ANY CHANGES TO THIS CLASS SHOULD BE MADE IN THOSE CLASSES AS WELL.
 */
public class Campaign {

	//---------------------------------------------------------------------------------------
	//---------------------------------- Constants ------------------------------------------
	//---------------------------------------------------------------------------------------	
	//Status constants for Campaign processing
	final public static String CREATED = "CREATED"; 
	final public static String READY = "READY"; 
	final public static String QUEUED = "QUEUED"; 
	final public static String PROCESSING = "PROCESSING"; 
	final public static String FINISHED = "FINISHED"; 
	final public static String ERROR = "ERROR"; 

	//Error constants (Note: String constants instead of Enum for symmetry with PHP/Javascript)
	final public static String SCREENSHOTCAPTURE = "SCREENSHOTCAPTURE"; 
	final public static String POWERPOINTGENERATION = "POWERPOINTGENERATION"; 
	final public static String CAMPAIGNEMAILSEND = "CAMPAIGNEMAILSEND"; 

	//URL path for PowerPoint files
	final public static String POWERPOINTURLPATH = "http://s3.amazonaws.com/" + ASRProperties.containerForPowerPoints() + "/";
	
	//---------------------------------------------------------------------------------------
	//--------------------------------- Static Methods --------------------------------------
	//---------------------------------------------------------------------------------------	
	//***************************** Public Static Methods ***********************************
	/**
	 * Retrieves and returns a Campaign from the database with the provided ID.
	 * 
	 * @param campaignID	ID of the Campaign
	 * @return				Campaign on success and NULL on failure
	 */	
	public static Campaign getCampaign(int campaignID) {
		
		//Check to see if a Campaign with that ID exists in the database
		//Redundant check along with constructor
		String getCampaignQuery = "SELECT * FROM campaigns WHERE CMP_id = " + campaignID;
		try (ResultSet campaignResults = ASRDatabase.executeQuery(getCampaignQuery)) {
	
			//If a match was found, return the Campaign 
			if (campaignResults.next()) {
				return new Campaign(campaignID);
			}	
			
			//Otherwise simply return null
			else {
				System.out.println("Unable to get Campaign with ID: " + campaignID);
				return null;
			}
		}
		
		//If we couldn't query the database correctly or another error occurred, print it out and return null
		catch (Exception e) {
			e.printStackTrace();;
			return null;
		}		
	}
	
	//---------------------------------------------------------------------------------------
	//------------------------------------ Variables ----------------------------------------
	//---------------------------------------------------------------------------------------	
	//******************************** Private Variables ************************************
	/**
	 * ID of the Campaign
	 */
	final private int _id;

	/**
	 * UUID of the Campaign
	 */
	final private String _uuid;
	
	/**
	 * Customer name for the campaign
	 */
	final private String _customerName;
	
	/**
	 * AdShots associated with the Campaign
	 */
	private Set<AdShot> _adShots;
	
	/**
	 * PowerPointBackground for the final PowerPoint
	 */
	private PowerPointBackground _powerPointBackground;
	
	/**
	 * Filename of the final PowerPoint
	 */
	private String _powerPointFilename;
	
	/**
	 * The current processing status of the Campaign.
	 * Possible values set by public static status members: CREATED, QUEUED, PROCESSING, FINISHED, ERROR
	 */
	private String _status;

	/**
	 * Error message if an error occurred while processing the Campaign
	 */
	private String _errorMessage;
	
	/**
	 * The timestamp the Campaign was inserted into the database
	 */
	private Timestamp _createdTimestamp;
	
	/**
	 * The timestamp the Campaign status was set to READY
	 */
	private Timestamp _readyTimestamp;
	
	/**
	 * The timestamp the Campaign status was set to QUEUED
	 */
	private Timestamp _queuedTimestamp;
	
	/**
	 * The UNIX timestamp the Campaign status was set to PROCESSING
	 */
	private Timestamp _processingTimestamp;
	
	/**
	 * The UNIX timestamp the Campaign status was set to FINISHED
	 */
	private Timestamp _finishedTimestamp;

	/**
	 * The UNIX timestamp the Campaign status was set to ERROR when an error occurred
	 */
	private Timestamp _errorTimestamp;
	
	//---------------------------------------------------------------------------------------
	//------------------------ Constructors/Copiers/Destructors -----------------------------
	//---------------------------------------------------------------------------------------
	private Campaign(int campaignID) {
		
		//Get the Campaign information from the database
		String getCampaignQuery = "SELECT * FROM campaigns WHERE CMP_id = " + campaignID;
		try (ResultSet campaignResult = ASRDatabase.executeQuery(getCampaignQuery)) {		
		
			//If a Creative was found with the passed ID, set the instance's members to its details
			if (campaignResult.next()) {
				
				_id = campaignResult.getInt("CMP_id");
				_uuid = campaignResult.getString("CMP_uuid");
				_customerName = campaignResult.getString("CMP_customerName");
				_powerPointFilename = campaignResult.getString("CMP_powerPointFilename");
				_status = campaignResult.getString("CMP_status");
				_errorMessage = campaignResult.getString("CMP_errorMessage");
				_createdTimestamp = campaignResult.getTimestamp("CMP_createdTimestamp");
				_readyTimestamp = campaignResult.getTimestamp("CMP_readyTimestamp");
				_queuedTimestamp = campaignResult.getTimestamp("CMP_queuedTimestamp");
				_processingTimestamp = campaignResult.getTimestamp("CMP_processingTimestamp");
				_finishedTimestamp = campaignResult.getTimestamp("CMP_finishedTimestamp");
				_errorTimestamp = campaignResult.getTimestamp("CMP_errorTimestamp");
			}	
			
			//Otherwise throw an error
			else {
				throw new AdShotRunnerException("Could not find Campaign with ID: " + campaignID);
			}
			
			//Get the PowerPointBackground for the Campaign
			_powerPointBackground = PowerPointBackground.getPowerPointBackground(campaignResult.getInt("CMP_PPB_id"));
			
		} catch (Exception e) {
			throw new AdShotRunnerException("Could not query database for Campaign with ID: " + campaignID, e);
		}				
		
		
		//Get the AdShots associated with the Campaign
		_adShots = new LinkedHashSet<AdShot>();
		String getAdShotsQuery = "SELECT * FROM adshots WHERE ADS_CMP_id = " + campaignID;
		try (ResultSet adshotsResult = ASRDatabase.executeQuery(getAdShotsQuery)) {
			while (adshotsResult.next()) {
				_adShots.add(AdShot.getAdShot(adshotsResult.getInt("ADS_id")));
			}
		} catch (Exception e) {
			throw new AdShotRunnerException("Could not query database for campaign AdShots: " + campaignID, e);
		}				
	}
	
	//---------------------------------------------------------------------------------------
	//------------------------------- Modification Methods ----------------------------------
	//---------------------------------------------------------------------------------------
	//********************************* Public Methods **************************************		
	/**
	 * Generates the final PowerPoint using the instance's AdShots, uploads it to storage,
	 * and stores its info in the database.
	 * 
	 * If there are no AdShots with a FINISHED status, an exception will be thrown.
	 * 
	 * @return		TRUE on success
	 */
	public boolean generatePowerPoint() throws Exception {
		
		//Verify at least one AdShot successfully finished. Throw an exception otherwise..
		boolean foundFinishedAdShot = false;
		for (AdShot currentAdShot : _adShots) {
			if (currentAdShot.status().equals(AdShot.FINISHED)) {foundFinishedAdShot = true;}
		}
		if (!foundFinishedAdShot) {throw new AdShotRunnerException("No finished AdShots available");}
		
		//Get the background. On failure, throw an exception.
		File backgroundImageFile = _powerPointBackground.file();
		if (backgroundImageFile == null) {throw new AdShotRunnerException("Could not download PowerPointBackground file");}
		
		//Get the current date as a MM/DD/YYYY string for the PowerPoint title page
		String campaignDate = new SimpleDateFormat("MM/dd/yyyy").format(Calendar.getInstance().getTime());
		
		//Create the PowerPoint
		CampaignPowerPoint powerPoint = new CampaignPowerPoint(_customerName, campaignDate, 
															   _powerPointBackground.fontColor(), 
															   backgroundImageFile);
		
		//Add a slide for each finished Desktop AdShot
		for (AdShot currentAdShot : _adShots) {
			if ((currentAdShot.status().equals(AdShot.FINISHED)) && (!currentAdShot.mobile())) {
				powerPoint.addSlide(currentAdShot);
			}
		}
		
		//Add a slide for each finished Mobile AdShot
		for (AdShot currentAdShot : _adShots) {
			if ((currentAdShot.status().equals(AdShot.FINISHED)) && (currentAdShot.mobile())) {
				powerPoint.addSlide(currentAdShot);
			}
		}
		
		//Create the PowerPoint filename
		String pptxCustomerName = _customerName.trim().replace(' ', '-');
		String pptxDate = new SimpleDateFormat("yyyy-MM-dd").format(Calendar.getInstance().getTime());
		String pptxTimestamp = Long.toString(Instant.now().getEpochSecond());
		pptxTimestamp = pptxTimestamp.substring(pptxTimestamp.length() - 6);
		String powerPointFilename = pptxCustomerName + "-" + pptxDate + "-" + pptxTimestamp + ".pptx";
		
		//Save the PowerPoint to a file and upload it
		powerPoint.save(ASRProperties.pathForTemporaryFiles() + powerPointFilename); 
		
    	try {Thread.sleep(250);} catch (InterruptedException e) {} //Make sure the file has time to save

		FileStorageClient.saveFile(ASRProperties.containerForPowerPoints(), 
				   				   ASRProperties.pathForTemporaryFiles() + powerPointFilename, powerPointFilename);
		
		//Store the new PowerPoint filename in the Campaign database entry
		ASRDatabase.executeUpdate("UPDATE campaigns " +
				  				  "SET CMP_powerPointFilename = '" + powerPointFilename + "' " +
				  				  "WHERE CMP_id = " + _id);
		
		//Store the filename in the instance
		_powerPointFilename = powerPointFilename;
		
		//Delete the files
		backgroundImageFile.delete();
		new File(ASRProperties.pathForTemporaryFiles() + powerPointFilename).delete();
		
		//Return true for success
		return true;
	}
	
	/**
	 * Returns the email address of the user who created the Campaign
	 * 
	 * @return 	Email address of the user who created the Campaign
	 */
	public String userEmailAddress() {
		
		//Query the database for the user
		try (ResultSet userResult = ASRDatabase.executeQuery("SELECT * " + 
															"FROM campaigns " + 
															"LEFT JOIN users ON CMP_USR_id = USR_id " +
															"WHERE CMP_id = " + _id)) {
			
			//Return the email address
			userResult.next();
			return userResult.getString("USR_emailAddress");
			
		} catch (Exception e) {
			e.printStackTrace();
	    	throw new AdShotRunnerException("Could not retrieve user email address for campgain: " + _id, e);
		}
	}
	
	/**
	 * Sets the processing status of the Campaign and sets the associated timestamp.
	 * 
	 * The options are the static members: QUEUED, PROCESSING, FINISHED
	 * 
	 * If a timestamp for the passed status already exists, this function will overwrite it
	 * with the new current time.
	 * 
	 * If the provided status is empty or does not exist, an exception is thrown 
	 * 
	 * The ERROR status cannot be set with this function. setError(...) should be used.
	 * 
	 * @param campaignStatus	Campaign status has defined by static members: QUEUED, PROCESSING, FINISHED
	 */
	public void setStatus(String campaignStatus) {
		
		//If the status is an empty string, throw an error
		if (campaignStatus.isEmpty()) {
        	throw new AdShotRunnerException("Empty string passed as a Campaign status");
		}
		
		//Determine the timestamp field to change based on the status.
		String timestampField = "";
		switch (campaignStatus) {
		
			case READY: timestampField = "CMP_readyTimestamp"; break;
			case QUEUED: timestampField = "CMP_queuedTimestamp"; break;
			case PROCESSING: timestampField = "CMP_processingTimestamp"; break;
			case FINISHED: timestampField = "CMP_finishedTimestamp"; break;
			
			//If the passed status did not match an official status, return an exception
			default: 
	        	throw new AdShotRunnerException("Attempt to set non-permitted Campaign status: " + campaignStatus);
		}
		
		//Update the status in the database
		try {
			ASRDatabase.executeUpdate("UPDATE campaigns " +
									  "SET CMP_status = '" + campaignStatus + "', " + 
									 	   timestampField + " = CURRENT_TIMESTAMP " +
									  "WHERE CMP_id = " + _id);
			
			//Set the new status in the instance
			_status = campaignStatus;			
		} catch (Exception e) {
			e.printStackTrace();
        	throw new AdShotRunnerException("Could not update Campaign status in the database: " + campaignStatus, e);
		}
		
		//Query the database for the new timestamp. This is to prevent localization errors.
		//Load all three timestamps to simplify code
		String getCampaignQuery = "SELECT * FROM campaigns WHERE CMP_id = " + _id;
		try (ResultSet campaignResult = ASRDatabase.executeQuery(getCampaignQuery)) {		
			if (campaignResult.next()) {
				_readyTimestamp = campaignResult.getTimestamp("CMP_readyTimestamp");
				_queuedTimestamp = campaignResult.getTimestamp("CMP_queuedTimestamp");
				_processingTimestamp = campaignResult.getTimestamp("CMP_processingTimestamp");
				_finishedTimestamp = campaignResult.getTimestamp("CMP_finishedTimestamp");
			}	
		} catch (Exception e) {
        	throw new AdShotRunnerException("Could not retrieve updated Campaign statuses ", e);
		}		

	}

	/**
	 * Sets the processing status of the Campaign to ERROR, sets the error timestamp, and stores
	 * the passed error message.
	 * 
	 * If an error already exists for the Campaign, this function will overwrite it and its
	 * timestamp.
	 * 
	 * @param errorMessage	Error message
	 */
	public void setError(String errorMessage) {
				
		//Update the status in the database
		try {
			ASRDatabase.executeUpdate("UPDATE campaigns " +
									  "SET CMP_status = '" + ERROR + "', " + 
								 	 	  "CMP_errorMessage = '" + errorMessage + "', " +
								 	 	  "CMP_errorTimestamp = CURRENT_TIMESTAMP " +
									  "WHERE CMP_id = " + _id);
			
			//Set the instance's error message and status
			_errorMessage = errorMessage;
			_status = ERROR;
			
		} catch (Exception e) {
        	throw new AdShotRunnerException("Could not store Campaign error in the database: " + errorMessage, e);
		}
		
		//Query the database for the new timestamp. This is to prevent localization errors.
		String getCampaignQuery = "SELECT * FROM campaigns WHERE CMP_id = " + _id;
		try (ResultSet campaignResult = ASRDatabase.executeQuery(getCampaignQuery)) {		
			if (campaignResult.next()) {
				_errorTimestamp = campaignResult.getTimestamp("CMP_errorTimestamp");
			}	
		} catch (Exception e) {
        	throw new AdShotRunnerException("Could not retrieve updated Campaign error status ", e);
		}		

	}
	
	//---------------------------------------------------------------------------------------
	//----------------------------------- Accessors -----------------------------------------
	//---------------------------------------------------------------------------------------
	//********************************* Public Accessors ************************************
	/**
	 * @return	ID of the Campaign
	 */
	public int id() {return _id;}
	
	/**
	 * @return	UUID of the Campaign
	 */
	public String uuid() {return _uuid;}
	
	/**
	 * @return	Customer name for the campaign
	 */
	public String customerName() {return _customerName;}
	
	/**
	 * @return	AdShots associated with the Campaign
	 */
	public Set<AdShot> adShots() {return _adShots;}
	
	/**
	 * @return	PowerPointBackground for the final PowerPoint
	 */
	public PowerPointBackground powerPointBackground() {return _powerPointBackground;}
	
	/**
	 * @return	Filename of the final PowerPoint
	 */
	public String powerPointFilename() {return _powerPointFilename;}
	
	/**
	 * @return	Full URL including protocol to final PowerPoint
	 */
	public String powerPointURL() {
		if (_powerPointFilename.isEmpty()) {return "";}
		return POWERPOINTURLPATH + _powerPointFilename;
	}
		
	/**
	 * @return	The current processing status of the Campaign. (Options static members: QUEUED, PROCESSING, FINISHED, ERROR)
	 */
	public String status() {return _status;}
	
	/**
	 * @return	Error message if an error occurred while processing the Campaign
	 */
	public String errorMessage() {return _errorMessage;}

	/**
	 * @return	The timestamp the Campaign was inserted into the database
	 */
	public Timestamp createdTimestamp() {return _createdTimestamp;}

	/**
	 * @return	The timestamp the Campaign status was set to READY
	 */
	public Timestamp readyTimestamp() {return _readyTimestamp;}

	/**
	 * @return	The timestamp the Campaign status was set to QUEUED
	 */
	public Timestamp queuedTimestamp() {return _queuedTimestamp;}

	/**
	 * @return	The timestamp the Campaign status was set to PROCESSING
	 */
	public Timestamp processingTimestamp() {return _processingTimestamp;}

	/**
	 * @return	The timestamp the Campaign status was set to FINISHED
	 */
	public Timestamp finishedTimestamp() {return _finishedTimestamp;}

	/**
	 * @return	The timestamp the Campaign status was set to ERROR when an error occurred
	 */
	public Timestamp errorTimestamp() {return _errorTimestamp;}

}

File: frontend/classes/AdShotRunner/Campaigns/Campaign.php
<?PHP
/**
* Contains the class for creating and retrieving Campaigns
*
* This class mirrors the Java and Javascript Creative classes
* ANY CHANGES TO THIS CLASS SHOULD BE MADE IN THOSE CLASSES AS WELL.
*
* @package Adshotrunner
* @subpackage Classes
*/

namespace AdShotRunner\Campaigns;

use AdShotRunner\Database\ASRDatabase;
use AdShotRunner\System\ASRProperties;
use AdShotRunner\Utilities\FileStorageClient;
use AdShotRunner\PowerPoint\PowerPointBackground;
use AdShotRunner\Campaigns\AdShot;

/**
* The Creative class creates creatives and retrieves them from the database. 
*
* This class mirrors the Java and Javascript Creative classes
* ANY CHANGES TO THIS CLASS SHOULD BE MADE IN THOSE CLASSES AS WELL.
*/
class Campaign {

	//---------------------------------------------------------------------------------------
	//---------------------------------- Constants ------------------------------------------
	//---------------------------------------------------------------------------------------
	//Status constants for Campaign processing
	const CREATED = "CREATED"; 
	const READY = "READY"; 
	const QUEUED = "QUEUED"; 
	const PROCESSING = "PROCESSING"; 
	const FINISHED = "FINISHED"; 
	const ERROR = "ERROR"; 

	//Error constants
	const SCREENSHOTCAPTURE = "SCREENSHOTCAPTURE"; 
	const POWERPOINTGENERATION = "POWERPOINTGENERATION"; 
	const CAMPAIGNEMAILSEND = "CAMPAIGNEMAILSEND"; 

	//URL path for screenshot images
	const POWERPOINTURLPATH = "https://s3.amazonaws.com/";

	//---------------------------------------------------------------------------------------
	//---------------------------------- Static Methods -------------------------------------
	//---------------------------------------------------------------------------------------	
	//******************************* Public Static Methods *********************************
	/**
	 * Retrieves and returns a Campaign from the database with the provided ID.
	 * 
	 * @param 	int 		campaignID	ID of the Campaign
	 * @return 	Campaign				Campaign on success and NULL on failure
	*/
	static public function getCampaign($campaignID) {

		//Verify the ID and that it's not less than 1. If not, return NULL.
		if ((!$campaignID) || (!is_numeric($campaignID)) || ($campaignID < 1)) {return NULL;}
		
		//Check to see if a Campaign with that ID exists in the database
		$getCampaignQuery = "SELECT * FROM campaigns WHERE CMP_id = $campaignID";
		$campaignResult = ASRDatabase::executeQuery($getCampaignQuery);
		$campaignDetails = $campaignResult->fetch_assoc();

		//If no creative was found, return NULL.
		if (!$campaignDetails) {return NULL;}

		//Otherwise, create a new Campaign and put the database information into it
		$retrievedCampaign = new Campaign();
		$retrievedCampaign->_id = $campaignID;
		$retrievedCampaign->_uuid = $campaignDetails['CMP_uuid'];
		$retrievedCampaign->_customerName = $campaignDetails['CMP_customerName'];
		$retrievedCampaign->_powerPointFilename = $campaignDetails['CMP_powerPointFilename'];
		$retrievedCampaign->_status = $campaignDetails['CMP_status'];
		$retrievedCampaign->_errorMessage = $campaignDetails['CMP_errorMessage'];
		$retrievedCampaign->_createdTimestamp = strtotime($campaignDetails['CMP_createdTimestamp']);
		$retrievedCampaign->_readyTimestamp = strtotime($campaignDetails['CMP_readyTimestamp']);
		$retrievedCampaign->_queuedTimestamp = strtotime($campaignDetails['CMP_queuedTimestamp']);
		$retrievedCampaign->_processingTimestamp = strtotime($campaignDetails['CMP_processingTimestamp']);
		$retrievedCampaign->_finishedTimestamp = strtotime($campaignDetails['CMP_finishedTimestamp']);
		$retrievedCampaign->_errorTimestamp = strtotime($campaignDetails['CMP_errorTimestamp']);

		//Get the PowerPointBackground
		$retrievedCampaign->_powerPointBackground = PowerPointBackground::getPowerPointBackground($campaignDetails['CMP_PPB_id']);

		//Get the AdShots associated with the Campaign
		$adShotsQuery = "SELECT * FROM adshots WHERE ADS_CMP_id = $campaignID";
		$adShotsResults = ASRDatabase::executeQuery($adShotsQuery);

		//Add the AdShots to the Campaign
		$retrievedCampaign->_adShots = [];
		while ($adShotEntry = $adShotsResults->fetch_assoc()) {
			$retrievedCampaign->_adShots[] = AdShot::getAdShot($adShotEntry['ADS_id']);
		}

		//Return the Campaign
		return $retrievedCampaign;
	}

	/**
	 * Retrieves and returns a Campaign from the database with the provided UUID.
	 * 
	 * @param 	String 		campaignUUID	UUID of the Campaign
	 * @return 	Campaign					Campaign on success and NULL on failure
	*/
	static public function getCampaignByUUID($campaignUUID) {

		//Verify the ID and that it's not an empty string. If not, return NULL.
		if ((!$campaignUUID) || ($campaignUUID == "")) {return NULL;}
		
		//Check to see if a Campaign with that UUID exists in the database
		$getCampaignQuery = "SELECT * FROM campaigns WHERE CMP_uuid = '$campaignUUID'";
		$campaignResult = ASRDatabase::executeQuery($getCampaignQuery);
		$campaignDetails = $campaignResult->fetch_assoc();

		//If no creative was found, return NULL.
		if (!$campaignDetails) {return NULL;}

		//Otherwise, get and return the full Campaign
		return Campaign::getCampaign($campaignDetails["CMP_id"]);
	}

	/**
	 * Creates a new Campaign using the passed information and returns it.
	 *
	 * The Campaign status is set to CREATED.
	 *
	 * @param 	String 		customerName				Customer name for the Campaign
	 * @param 	int 		powerPointBackgroundID		ID of the PowerPointBackground to use for final PowerPoint
	 * @return 	Campaign								Newly created Campaign
	*/
	static public function create($customerName, $powerPointBackgroundID) {

		//Verify a customer name was passed. If not, return NULL.
		if ((!$customerName) || ($customerName == "")) {return NULL;}

		//Verify the PowerPointBackground exists
		$campaignBackground = PowerPointBackground::getPowerPointBackground($powerPointBackgroundID);
		if (!$campaignBackground) {return NULL;}

		//Create a unique UUID for the Campaign 
		$campaignUUID = self::getUUID();

		//Add the Campaign to the campaigns table
		$addCampaignQuery = "INSERT INTO campaigns (CMP_uuid,
												    CMP_customerName,
												    CMP_PPB_id,
												    CMP_USR_id)
						 	 VALUES ('" . ASRDatabase::escape($campaignUUID) . "',
									 '" . ASRDatabase::escape($customerName) . "',
									 						  $powerPointBackgroundID,
									 '" . USERID . "')";
		ASRDatabase::executeQuery($addCampaignQuery);
		
		//Get the ID of the new Campaign
		$campaignID = ASRDatabase::lastInsertID();

		//Get the newly created Campaign with all of its defaults such as CREATED timestamp set
		return self::getCampaign($campaignID);
	}

	//******************************* Private Static Methods *********************************
	/**
	 * Returns a random UUID
	 *
	 * Taken from: https://stackoverflow.com/a/15875555
	 * 
	 * @return 	String		Unique random-generated UUID
	*/
	static private function getUUID() {

		$data = random_bytes(16);
	    $data[6] = chr(ord($data[6]) & 0x0f | 0x40); // set version to 0100
	    $data[8] = chr(ord($data[8]) & 0x3f | 0x80); // set bits 6-7 to 10
	    return vsprintf('%s%s-%s-%s-%s-%s%s%s', str_split(bin2hex($data), 4));
    }

	//---------------------------------------------------------------------------------------
	//------------------------------------ Variables ----------------------------------------
	//---------------------------------------------------------------------------------------	
	//******************************** Private Variables ************************************
	/**
	* @var int  ID of the Campaign
	*/
	private $_id;

	/**
	* @var string  UUID of the Campaign
	*/
	private $_uuid;

	/**
	* @var string  Customer name for the Campaign
	*/
	private $_customerName;

	/**
	* @var array  AdShots associated with the Campaign (array)
	*/
	private $_adShots;

	/**
	* @var PowerPointBackground  PowerPointBackground for the final PowerPoint
	*/
	private $_powerPointBackground;

	/**
	* @var string  Filename of the final PowerPoint
	*/
	private $_powerPointFilename;

	/**
	* @var string  Current processing status of the Campaign (Static class constants: CREATED, QUEUED, PROCESSING, FINISHED, ERROR)
	*/
	private $_status;

	/**
	* @var string  Error message if an error occurred while processing the Campaign
	*/
	private $_errorMessage;

	/**
	* @var int  The UNIX timestamp the Campaign was inserted into the database
	*/
	private $_createdTimestamp;

	/**
	* @var int  The UNIX timestamp the Campaign status was set to READY
	*/
	private $_readyTimestamp;

	/**
	* @var int  The UNIX timestamp the Campaign status was set to QUEUED
	*/
	private $_queuedTimestamp;

	/**
	* @var int  The UNIX timestamp the Campaign status was set to PROCESSING
	*/
	private $_processingTimestamp;

	/**
	* @var int  The UNIX timestamp the Campaign status was set to FINISHED
	*/
	private $_finishedTimestamp;

	/**
	* @var int  The UNIX timestamp the Campaign status was set to ERROR when an error occurred
	*/
	private $_errorTimestamp;


	//---------------------------------------------------------------------------------------
	//------------------------ Constructors/Copiers/Destructors -----------------------------
	//---------------------------------------------------------------------------------------
	/**
	* Private constructor for static factories
	*/
	private function __construct() {}


	//---------------------------------------------------------------------------------------
	//------------------------------- Modification Methods ----------------------------------
	//---------------------------------------------------------------------------------------
	//********************************* Public Methods **************************************
	/**
	 * Creates a new AdShot associated to the Campaign.
	 *
	 * @param 	String 		requestedURL		Requested URL of the AdShot (Before StoryFinder)
	 * @param 	boolean 	useStoryFinder		TRUE to use a story from the requested URL page
	 * @param 	boolean 	useMobile			TRUE to take a mobile screenshot
	 * @param 	boolean 	useBelowTheFold		TRUE to take a Below-the-Fold screenshot
	 * @return 	AdShot					Newly created AdShot
	*/
	public function createAdShot($requestedURL, $useStoryFinder, $useMobile, $useBelowTheFold) {

		//Create the AdShot. Return NULL if nothing returned.
		$newAdShot = AdShot::create($this->id(), $requestedURL, $useStoryFinder, $useMobile, $useBelowTheFold);
		if (!$newAdShot) {return NULL;}

		//Add the AdShot to the instance
		$this->_adShots[] = $newAdShot;

		//Return the AdShot
		return $newAdShot;
	}

	/**
	 * Creates a new AdShot associated to the Campaign using the provided JSON string,
	 * which is generated by a javascript AdShotBuilder on the client.
	 *
	 * @param 	String 		adShotJSON		New AdShot JSON string generated by an AdShotBuilder
	 * @return 	AdShot						Newly created AdShot
	*/
	public function createAdShotFromJSON($adShotJSON) {

		//Decode the JSON and validate the variables
		$jsonObject = json_decode($adShotJSON, true);
		if (($jsonObject["requestedURL"] == null)) {return null;}

		//Create the AdShot. Return NULL if nothing returned.
		$newAdShot = AdShot::create($this->id(), $jsonObject["requestedURL"], $jsonObject["storyFinder"], 
												 $jsonObject["mobile"], $jsonObject["belowTheFold"]);
		if (!$newAdShot) {return NULL;}

		//Add the Creative to the AdShot
		foreach ($jsonObject["creativeIDs"] as $currentID) {
			$addCreativeSuccess = $newAdShot->addCreativeByID($currentID);
			if (!$addCreativeSuccess) {return NULL;}
		}

		//Add the AdShot to the instance
		$this->_adShots[] = $newAdShot;

		//Return the AdShot
		return $newAdShot;
	}

	/**
	 * Marks the Campaign ready for processing.
	 *
	 * This notifies the processing server the Campaign is fully built and is ready for its
	 * AdShots to be imaged and the resulting PowerPoint to be made.
	 * 
	 * @return 	boolean		True on success and NULL on failure
	 */
	public function readyForProcessing() {

		//Update the status in the database
		$setStatusQuery = "UPDATE campaigns 
						   SET CMP_status = '" . $this::READY . "', 
						   	   CMP_readyTimestamp = CURRENT_TIMESTAMP 
						   WHERE CMP_id = " . $this->_id;
		ASRDatabase::executeQuery($setStatusQuery);

		//Set the new status in the instance
		$this->_status = $this::QUEUED;
		
		//Query the database for the new timestamp. This is to prevent localization errors.
		$getCampaignQuery = "SELECT * FROM campaigns WHERE CMP_id = " . $this->_id;
		$campaignResult = ASRDatabase::executeQuery($getCampaignQuery);
		$campaignEntry = $campaignResult->fetch_assoc();
		$this->_readyTimestamp = strtotime($campaignEntry["CMP_readyTimestamp"]);

		//Return success
		return TRUE;
	}

	/**
	 * Returns a JSON object representing all of the member values of the Campaign instance.
	 *
	 * The final object also includes full URLs for the image and tag page files.
	 * 
	 * @return 	String		JSON object representing the member values of the instance
	*/
	public function toJSON() {

		//Create the associative array with all the key value pairs
		$jsonObject = [];
		$jsonObject["id"] = $this->_id;
		$jsonObject["uuid"] = $this->_uuid;
		$jsonObject["customerName"] = $this->_customerName;
		$jsonObject["powerPointFilename"] = $this->_powerPointFilename;
		$jsonObject["powerPointURL"] = $this->powerPointURL();
		$jsonObject["status"] = $this->_status;
		$jsonObject["errorMessage"] = $this->_errorMessage;
		$jsonObject["createdTimestamp"] = $this->_createdTimestamp;
		$jsonObject["readyTimestamp"] = $this->_readyTimestamp;
		$jsonObject["queuedTimestamp"] = $this->_queuedTimestamp;
		$jsonObject["processingTimestamp"] = $this->_processingTimestamp;
		$jsonObject["finishedTimestamp"] = $this->_finishedTimestamp;
		$jsonObject["errorTimestamp"] = $this->_errorTimestamp;

		//Add the AdShots as an array of their own JSON strings
		$jsonObject["adShots"] = [];
		foreach ($this->_adShots as $currentAdShot) {
			$jsonObject["adShots"][] = $currentAdShot->toJSON();
		}

		//Convert the array to a JSON string and return it
		return json_encode($jsonObject);
	}	

	//---------------------------------------------------------------------------------------
	//----------------------------------- Accessors -----------------------------------------
	//---------------------------------------------------------------------------------------
	//********************************* Public Accessors ************************************
	/**
	* @return int	ID of the Campaign
	*/
	public function id() {
		return $this->_id;
	}

	/**
	* @return string	UUID of the Campaign
	*/
	public function uuid() {
		return $this->_uuid;
	}

	/**
	* @return string	Customer name for the Campaign
	*/
	public function customerName() {
		return $this->_customerName;
	}

	/**
	* @return array 	AdShots associated with the Campaign (array)
	*/
	public function adShots() {
		return $this->_adShots;
	}

	/**
	* @return PowerPointBackground 	PowerPointBackground for the final PowerPoint
	*/
	public function powerPointBackground() {
		return $this->_powerPointBackground;
	}

	/**
	* @return string	Filename of the final PowerPoint
	*/
	public function powerPointFilename() {
		return $this->_powerPointFilename;
	}

	/**
	* @return string	Full URL including protocol to the final PowerPoint
	*/
	public function powerPointURL() {
		if ($this->_powerPointFilename == "") {return "";}
		return $this::POWERPOINTURLPATH . ASRProperties::containerForPowerPoints() . "/" . $this->_powerPointFilename;
	}

	/**
	* @return string	Current processing status of the Campaign. (Options const members: CREATED, QUEUED, PROCESSING, FINISHED, ERROR)
	*/
	public function status() {
		return $this->_status;
	}

	/**
	* @return string	Error message if an error occurred while processing the Campaign
	*/
	public function errorMessage() {
		return $this->_errorMessage;
	}

	/**
	* @return int	UNIX Timestamp the Campaign was inserted into the database
	*/
	public function createdTimestamp() {
		return $this->_createdTimestamp;
	}

	/**
	* @return int	UNIX Timestamp the Campaign status was set to READY
	*/
	public function readyTimestamp() {
		return $this->_readyTimestamp;
	}

	/**
	* @return int	UNIX Timestamp the Campaign status was set to QUEUED
	*/
	public function queuedTimestamp() {
		return $this->_queuedTimestamp;
	}

	/**
	* @return int	UNIX Timestamp the Campaign status was set to PROCESSING
	*/
	public function processingTimestamp() {
		return $this->_processingTimestamp;
	}

	/**
	* @return int	UNIX Timestamp the Campaign status was set to FINISHED
	*/
	public function finishedTimestamp() {
		return $this->_finishedTimestamp;
	}

	/**
	* @return int	UNIX timestamp the Campaign status was set to ERROR when an error occurred
	*/
	public function errorTimestamp() {
		return $this->_errorTimestamp;
	}




}